### ISSUE TITLE
Make the on-screen keys optional

### ISSUE THREAD

**Issue by sspanak** [sspanak/tt9#104]
All functionality is available using the hardware keys, so they don't need to be displayed. 

The only reason they are not optional is because Back does not work properly as Backspace in all applications, so there must be an alternative to the hardware key. However, after completing #35 and #29, it would be possible to customize the hardware key and the on-screen one will be completely obsolete.

For the reasons above:
- Add a "Show on-screen controls" checkbox on the Settings screen. The option controls whether the soft keys are displayed or not.
- Change the title of the option if it does not fit well on the screen.

After #35

**Comment by puffertron** [comment]
Would it be possible to have the on-screen keys mapped to the builtin menu that some flip phones have on the bottom of every screen? I'm, not sure what these widget menu things are called in code, but here's an example image of how my default keyboard uses these menus for character set selection, settings, and emojis. 
![screen](https://user-images.githubusercontent.com/27284164/202013576-9e5ad70a-6ed1-4490-a518-cd3aade7f5e1.png)

**Comment by sspanak** [comment]
These are definitely not standard Android widgets. Probably your phone manifacturer has chosen to add these for convenience when you use its default IME. 

I really doubt it is possible to wire Traditional T9 to them. And if it is, it would have to be a custom version that would work only on your phone.

Instead, I would recommend disabling the widgets, if possible. If not, just wait for the next release, where you would be able to disable the Traditional T9 on-screen buttons and use only your phone's native ones.

**Comment by puffertron** [comment]
Hm, These widgets are not only for the keyboard, almost every app I open has them on the bottom of the screen, and usually, at least one of the spaces is filled with the entry 'Menu' which does nothing when pressed on most apps. 

For some more context, the widget is not touch, it's mapped to the four 'function keys' on the left and right side of my dpad, and the centre button is my OK button on the dpad. The menu does work in the preinstalled apps on my phone so I have a hunch it's something similar to the (probably deprecated) options/menu button functionality that many phones had in the past. I also don't think it's possible to disable this widget :/

My phone is a Kyocera KYF31 on android 5.1, but I know another person who uses a Japanese flip-phone of a different brand and his phone uses these menus as well. Anyways, this isn't a priority- I'll look into it on my own time and if I find a way to use it maybe ill contribute! Thanks!

**Comment by sspanak** [comment]
I see. It sounds they are just labels on the screen that indicate what the keys do. And if we are talking about hardware keys, it is not so difficult to map them to TT9 functions. Since the next release, all key shortcuts will be customizable. Currently, I have added support for CALL, BACK, MENU, # and *, but adding any other keys is trivial.

Here is how it will look:
![Screenshot_20221117-145342](https://user-images.githubusercontent.com/27846105/202452625-1534be90-e8d0-4ac6-98ff-05a17adb4dc9.png)

If, even after that, you still find the settings are not flexible enough, just open a new issue and I should be able to improve them with a little help from your side (since the phone is in your hands).

**Comment by puffertron** [comment]
> I see. It sounds they are just labels on the screen that indicate what the keys do.
Yeah, that sounds about right. I found a way to hide them too.

That settings menu looks great! The keys that I described earlier aren't common, but here are keycodes that will probably be consistent across most Japanese phones and you could code them in and label them as Function 1 through 4:

Func 1: 131 (top left)
Func 2: 132 (top right)
Func 3: 133 (bottom left)
Func 4: 134 (bottom right)

Another option for the future could be to have a 'custom' option and prompt users to press a key to bind it, though that would take longer to implement.

> If, even after that, you still find the settings are not flexible enough, just open a new issue and I should be able to improve them with a little help from your side (since the phone is in your hands).

Will do! My flipphone-owning friend and I are very grateful. I'll have him do tests on his phone as well so we can get a better idea about compatibility.

**Comment by sspanak** [comment]
> Another option for the future could be to have a 'custom' option and prompt users to press a key to bind it, though that would take longer to implement.

I considered this too, but it seemed a bit more challenging, so I decided not to go this way for now.

> Func 1: 131 (top left)
Func 2: 132 (top right)
Func 3: 133 (bottom left)
Func 4: 134 (bottom right)

According to the [official manual](https://developer.android.com/reference/android/view/KeyEvent#KEYCODE_F1), these are recognized as F1, F2, F3, F4 (like on a computer keyboard). I have already added support for F1 and F2 and I can easily add F3 and F4, as well. However, I am unsure if I did it properly, because I don't have these keys on my phone. I will need you two to check if they appear as possible options when you go to the Settings screen.

I can build a beta verion `.apk` for you test. Or, if you prefer, build it on your own from the `new-settings-screen` branch. Have in mind you may need to uninstall and reinstall the app, if your phone sees the `.apk` as incompatible. This will erase all your settings and custom dictionary words (if any).

How would you like to proceed?

**Comment by puffertron** [comment]
A beta apk build sounds great! I'll probably end up using the f3 and f4 keys, so if you add them to the beta I can test everything at once.

I don't have any custom words saved, so I'm fine to uninstall and reinstall. Let me know when the release is up and I'll try it out!

**Comment by sspanak** [comment]
I came across this problem, but it doesn't happen every time or it happens under very specific circumstances. It may fail when you try to open the settings directly from the Android Settings, but I am not sure.

Could you please try again while you are in another application where you can type?

**Comment by puffertron** [comment]
Ah oops I forgot that I could just hold star to access settings. The remapping seems to be working perfectly, aside from that settings bug it's all looking good. Thanks!

**Comment by sspanak** [comment]
> Ah oops I forgot that I could just hold star to access settings. The remapping seems to be working perfectly, aside from that settings bug it's all looking good. Thanks!

So, I take F1-F4 buttons are working fine and I don't need to work on them anymore. That's great!

> One thing I am noticing is that my back button functions strangely when tt9 is enabled.

Now that you mentioned it, I think there may be some leftover code that could tamper with the button function. I'll double check.

Thank you very much for the feedback, it is most useful!

**Comment by puffertron** [comment]
That seems to work! no more issues with the back button after a quick test.

I am having another problem though, with loading the English dictionary. (Since the last beta, I didn't test for very long and I figured i was cancelling the loading accidentally by letting my phone fall asleep or exiting the menu screen)

Loading a dictionary takes a while, but I can't tell if it's done or not after trying to load it. Prediction mode doesn't show any suggestions with real words, it just shows groups of letters based off the keys I've pressed. I think prediction is working because when I add a word manually it shows up in the prediction bar.

Update: while writing this my phone has been trying to load the dictionary for more than 20 minutes now, I think there may be a bug here too.

**Comment by sspanak** [comment]
> That seems to work! no more issues with the back button after a quick test.

Great! I'll give it a couple more days and if don't find anything that requires code changes, we can pronounce the next release ready!

> I am having another problem though, with loading the English dictionary.

That's an unrelated problem. Let's continue the discussion [here](https://github.com/sspanak/tt9/issues/117).
